File: src/main/java/net/atos/client/zgw/zrc/model/RolOrganisatorischeEenheid.java
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2023 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.client.zgw.zrc.model;

import java.net.URI;
import java.util.Objects;
import java.util.UUID;

import org.apache.commons.lang3.StringUtils;

import nl.info.client.zgw.zrc.model.generated.BetrokkeneTypeEnum;
import nl.info.client.zgw.zrc.model.generated.OrganisatorischeEenheidIdentificatie;
import nl.info.client.zgw.ztc.model.generated.RolType;

/**
 * Manually copied from {@link nl.info.client.zgw.zrc.model.generated.RolOrganisatorischeEenheid} and modified to allow for
 * polymorphism using a generic base {@link Rol} class.
 * Ideally we would use the generated class, but currently we cannot get the OpenAPI Generator framework to generate
 * polymorphic relationships correctly.
 */
public class RolOrganisatorischeEenheid extends Rol<OrganisatorischeEenheidIdentificatie> {

    public RolOrganisatorischeEenheid() {
    }

    /**
     * For testing purposes only where we need a UUID.
     */
    public RolOrganisatorischeEenheid(
            final UUID uuid,
            final RolType roltype,
            final String roltoelichting,
            final OrganisatorischeEenheidIdentificatie betrokkeneIdentificatie
    ) {
        super(uuid, roltype, BetrokkeneTypeEnum.ORGANISATORISCHE_EENHEID, betrokkeneIdentificatie, roltoelichting);
    }

    public RolOrganisatorischeEenheid(
            final URI zaak,
            final RolType roltype,
            final String roltoelichting,
            final OrganisatorischeEenheidIdentificatie organisatorischeEenheid
    ) {
        super(zaak, roltype, BetrokkeneTypeEnum.ORGANISATORISCHE_EENHEID, organisatorischeEenheid, roltoelichting);
    }

    @Override
    protected boolean equalBetrokkeneIdentificatie(final OrganisatorischeEenheidIdentificatie identificatie) {
        final OrganisatorischeEenheidIdentificatie betrokkeneIdentificatie = getBetrokkeneIdentificatie();
        if (betrokkeneIdentificatie == identificatie) {
            return true;
        }
        if (identificatie == null) {
            return false;
        }
        return Objects.equals(betrokkeneIdentificatie.getIdentificatie(), identificatie.getIdentificatie());
    }

    @Override
    public String getNaam() {
        if (getBetrokkeneIdentificatie() == null) {
            return null;
        }
        return StringUtils.isNotEmpty(getBetrokkeneIdentificatie().getNaam()) ? getBetrokkeneIdentificatie().getNaam() :
                getIdentificatienummer();
    }

    @Override
    public String getIdentificatienummer() {
        if (getBetrokkeneIdentificatie() == null) {
            return null;
        }
        return getBetrokkeneIdentificatie().getIdentificatie();
    }

    @Override
    protected int hashCodeBetrokkeneIdentificatie() {
        return Objects.hash(getBetrokkeneIdentificatie().getIdentificatie());
    }
}

File: src/main/kotlin/nl/info/client/zgw/zrc/model/RolOrganisatorischeEenheid.kt
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2023, 2026 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.client.zgw.zrc.model

import nl.info.client.zgw.zrc.model.generated.BetrokkeneTypeEnum
import nl.info.client.zgw.zrc.model.generated.OrganisatorischeEenheidIdentificatie
import nl.info.client.zgw.ztc.model.generated.RolType
import org.apache.commons.lang3.StringUtils
import java.net.URI
import java.util.Objects
import java.util.UUID

/**
 * Manually copied from [nl.info.client.zgw.zrc.model.generated.RolOrganisatorischeEenheid] and modified to allow for
 * polymorphism using a generic base [Rol] class.
 * Ideally we would use the generated class, but currently we cannot get the OpenAPI Generator framework to generate
 * polymorphic relationships correctly.
 */
class RolOrganisatorischeEenheid : Rol<OrganisatorischeEenheidIdentificatie> {
    constructor() : super()

    /**
     * For testing purposes only where we need a UUID.
     */
    constructor(
        uuid: UUID,
        roltype: RolType,
        roltoelichting: String,
        betrokkeneIdentificatie: OrganisatorischeEenheidIdentificatie?
    ) : super(uuid, roltype, BetrokkeneTypeEnum.ORGANISATORISCHE_EENHEID, betrokkeneIdentificatie, roltoelichting)

    constructor(
        zaak: URI,
        roltype: RolType,
        roltoelichting: String,
        organisatorischeEenheid: OrganisatorischeEenheidIdentificatie?
    ) : super(zaak, roltype, BetrokkeneTypeEnum.ORGANISATORISCHE_EENHEID, organisatorischeEenheid, roltoelichting)

    @Suppress("ReturnCount")
    override fun equalBetrokkeneIdentificatie(identificatie: OrganisatorischeEenheidIdentificatie?): Boolean {
        val ownIdentificatie = betrokkeneIdentificatie
        if (ownIdentificatie === identificatie) {
            return true
        }
        if (identificatie == null) {
            return false
        }
        return Objects.equals(ownIdentificatie?.identificatie, identificatie.identificatie)
    }

    override val naam: String?
        get() {
            val identificatie = betrokkeneIdentificatie ?: return null
            return if (StringUtils.isNotEmpty(identificatie.naam)) identificatie.naam else identificatienummer
        }

    override val identificatienummer: String?
        get() = betrokkeneIdentificatie?.identificatie

    override fun hashCodeBetrokkeneIdentificatie(): Int = Objects.hash(betrokkeneIdentificatie?.identificatie)
}
